tools/snap/build_remote.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+import argparse
+import glob
+import datetime
+from multiprocessing import Pool, Process, Manager
+import re
+import subprocess
+import sys
+import time
+from os.path import join, realpath, dirname, basename
+
+
+CERTBOT_DIR = dirname(dirname(dirname(realpath(__file__))))
+PLUGINS = [basename(path) for path in glob.glob(join(CERTBOT_DIR, 'certbot-dns-*'))]
+
+
+def _execute_build(target, archs, status, workspace):
+    process = subprocess.Popen([
+        'snapcraft', 'remote-build', '--launchpad-accept-public-upload', '--recover', '--build-on', ','.join(archs)
+    ], stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, cwd=workspace)
+
+    line = process.stdout.readline()
+    while line:
+        _extract_state(target, line, status)
+        line = process.stdout.readline()
+
+
+def _build_snap(target, archs, status):
+    status[target] = {arch: '...' for arch in archs}
+
+    if target == 'certbot':
+        workspace = CERTBOT_DIR
+    else:
+        workspace = join(CERTBOT_DIR, target)
+        subprocess.check_output(
+            ('"{0}" tools/strip_hashes.py letsencrypt-auto-source/pieces/dependency-requirements.txt '
+             '| grep -v python-augeas > "{1}/snap-constraints.txt"').format(sys.executable, workspace),
+            shell=True, cwd=CERTBOT_DIR)
+
+    retry = 3
+    while retry:
+        _execute_build(target, archs, status, workspace)
+        if not [status[target] for arch in archs if status[target][arch] == '...']:
+            break
+
+        retry = retry - 1
+
+    return {target: workspace}
+
+
+def _extract_state(project, output, status):
+    match = re.match(r'^.*arch=(\w+)\s+state=([\w ]+).*$', output)
+    if match:
+        arch = match.group(1)
+        state = status[project]
+        state[arch] = match.group(2)
+
+        status[project] = state
+
+
+def _dump_status(archs, status, final=False):
+    while True:
+        if final:
+            print('Results for remote build finished at {0}'.format(datetime.datetime.now()))
+        else:
+            print('Remote build status at {0}'.format(datetime.datetime.now()))
+        print(' project                    {0}'.format(''.join('| {0}                       '.format(arch)
```

Comment:
What do you think about grabbing a third party library for this like [tabulate](https://pypi.org/project/tabulate/)? We could pin it in `dev_constraints.txt` and then just use `tools/pip_install.py` to install it before running this script. We could also automate this setup in `tox` (and make use of [positional arguments](https://tox.readthedocs.io/en/latest/example/general.html#interactively-passing-positional-arguments) for controlling the script) or create a package for this although the latter feels like overkill to me.

Alternatively, I think we could simplify things a bit and make them more flexible by making use of [format options](https://docs.python.org/3.8/library/string.html#formatspec) and maybe use [f-strings](https://www.python.org/dev/peps/pep-0498/) to easily make variables like the cell width dynamic.

I'd personally prefer to just use an existing library for this though.

Reply:
Well I would like to avoid if possible to depend on third party libraries. As you can see currently I am using only core Python modules in the script. This way the script can just be started without installing a development environment.

If you agree, I will look at the format options in this case.

Reply:
> This way the script can just be started without installing a development environment.

I agree that's a nice property in theory, but wonder how much it really buys us in practice. If a dev wants to run it locally, I think it's reasonable to require and expect them to already have the development environment set up. For CI, I think the performance overhead of running `tools/pip_install.py tabulate` is very minimal. You wouldn't have to even create a venv if you want even though it only takes a couple seconds to do so. In both cases, I think the import error that would be raised automatically if `tabulate` doesn't exist is pretty explanatory.

Did you have other use cases in mind for this script where not having any third party dependencies is a significant benefit? If not, I'd prefer to add a relatively popular dependency that allows us to make arbitrary, nicely formatted tables in just a couple lines of code.

If you do have a use case or you're not convinced in general, feel free to take a look at format options. The main thing I was thinking there is I think snippets like:
```
{0}'.format(states[arch] + ' ' * (27 - len(states[arch]))) for arch in archs)))
```
can become:
```
{0:<27}'.format(states[arch] for arch in archs)))
```
You could also maybe look into setting the cell width in a variable and reusing it everywhere, possibly by making use of [f-strings](https://www.python.org/dev/peps/pep-0498/) to simplify things, but that may be more trouble than it is worth.

Reply:
I tried an approach with f-string + format options and pushed it here: is it ok enough for you or do you still want to use tabulate?

Reply:
Thanks!

That looks simple enough to me that I think it's probably fine to leave it as is.